### PR TITLE
FileInfoExtractor: decode non-UTF-8 characters #182 #183

### LINKED ISSUE #182: FileInfoAnalyzer#isReused throws MalformedInputException while generating report for CS2103JAN2018-T11-B1/main
```
INFO: Extracting commits info for CS2103JAN2018-T11-B1/main...
Jul 12, 2018 5:06:40 PM reposense.authorship.FileInfoExtractor extractFileInfos
INFO: Extracting relevant file infos...
Jul 12, 2018 5:07:59 PM reposense.authorship.FileInfoAnalyzer isReused
WARNING: Input length = 1
java.nio.charset.MalformedInputException: Input length = 1
	at java.nio.charset.CoderResult.throwException(Unknown Source)
	at sun.nio.cs.StreamDecoder.implRead(Unknown Source)
	at sun.nio.cs.StreamDecoder.read(Unknown Source)
	at java.io.InputStreamReader.read(Unknown Source)
	at java.io.BufferedReader.read1(Unknown Source)
	at java.io.BufferedReader.read(Unknown Source)
	at java.io.BufferedReader.fill(Unknown Source)
	at java.io.BufferedReader.readLine(Unknown Source)
	at java.io.BufferedReader.readLine(Unknown Source)
	at reposense.authorship.FileInfoAnalyzer.isReused(FileInfoAnalyzer.java:102)
	at reposense.authorship.FileInfoAnalyzer.analyzeFile(FileInfoAnalyzer.java:40)
	at reposense.authorship.AuthorshipReporter.lambda$generateAuthorshipSummary$0(AuthorshipReporter.java:24)
	at java.util.stream.ReferencePipeline$3$1.accept(Unknown Source)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(Unknown Source)
	at java.util.stream.AbstractPipeline.copyInto(Unknown Source)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source)
	at java.util.stream.AbstractPipeline.evaluate(Unknown Source)
	at java.util.stream.ReferencePipeline.collect(Unknown Source)
	at reposense.authorship.AuthorshipReporter.generateAuthorshipSummary(AuthorshipReporter.java:26)
	at reposense.report.ReportGenerator.generateReposReport(ReportGenerator.java:45)
	at reposense.RepoSense.main(RepoSense.java:31)

### LINKED ISSUE #183: FileInfoExtractor#generateFileInfo throws MalformedInputException while generating report for CS2103JAN2018-F11-B4/main
```
INFO: Extracting commits info for CS2103JAN2018-F11-B4/main...
Jul 12, 2018 11:38:55 AM reposense.authorship.FileInfoExtractor extractFileInfos
INFO: Extracting relevant file infos...
Jul 12, 2018 11:38:55 AM reposense.authorship.FileInfoExtractor generateFileInfo
SEVERE: Input length = 2
java.nio.charset.MalformedInputException: Input length = 2
	at java.nio.charset.CoderResult.throwException(Unknown Source)
	at sun.nio.cs.StreamDecoder.implRead(Unknown Source)
	at sun.nio.cs.StreamDecoder.read(Unknown Source)
	at java.io.InputStreamReader.read(Unknown Source)
	at java.io.BufferedReader.read1(Unknown Source)
	at java.io.BufferedReader.read(Unknown Source)
	at java.io.BufferedReader.fill(Unknown Source)
	at java.io.BufferedReader.readLine(Unknown Source)
	at java.io.BufferedReader.readLine(Unknown Source)
	at reposense.authorship.FileInfoExtractor.generateFileInfo(FileInfoExtractor.java:80)
	at reposense.authorship.FileInfoExtractor.getAllFileInfo(FileInfoExtractor.java:62)
	at reposense.authorship.FileInfoExtractor.getAllFileInfo(FileInfoExtractor.java:58)
	at reposense.authorship.FileInfoExtractor.getAllFileInfo(FileInfoExtractor.java:58)
	at reposense.authorship.FileInfoExtractor.extractFileInfos(FileInfoExtractor.java:38)
	at reposense.authorship.AuthorshipReporter.generateAuthorshipSummary(AuthorshipReporter.java:21)
	at reposense.report.ReportGenerator.generateReposReport(ReportGenerator.java:45)
	at reposense.RepoSense.main(RepoSense.java:31)
```

### PR DESCRIPTION
**Pull request by yamidark**
Fixes #182  
Fixes #183 
```
FileInfoExtractor#generateFileInfo(String, String)
and FileInfoAnalyzer#isReused(String,String) use
Files#newBufferedReader(Path) to construct the BufferedReader for
reading of files.

This constructed BufferedReader uses the default UTF-8 Charset[1]
and a default CharsetDecoder[2] to decode the bytes
read from files into Strings.[3]

As some files being analyzed may contain special characters that are
not from the UTF-8 character set[3], it causes the BufferedReader to
throw a java.nio.charset.MalformedInputException.

As a result, it creates an incomplete FileInfo in
FileInfoExtractor#generateFileInfo(String, String) process, which causes the
program to crash when analysing the file in 
FileInfoAnalyzer#aggregateBlameAuthorInfo(RepoConfiguration, FileInfo).
This is due to the number of lines analyzed in git blame being greater than
the number of lines inside FileInfo. It will also cause the analysis of
some white-listed files to be skipped inside
FileInfoAnalyzer#isReused(String,String).

Let's update FileInfoExtractor#generateFileInfo(String, String)
and FileInfoAnalyzer#isReused(String,String) to construct the
BufferedReader with an java.io.FileReader instead.

The java.io.FileReader extends java.io.InputStreamReader,
which use the same default UTF-8 Charset. However, it will call 
StreamDecoder#forInputStreamReader(InputStream, Object, CharsetDecoder)
, which will set the CharsetDecoder to replace the malformed input
instead of throwing an error[5], thus allowing the FileInfo to be
constructed with the correct number of lines.

[1] Javadocs of Charset:
https://docs.oracle.com/javase/8/docs/api/java/nio/charset/Charset.html

[2] Javadocs of CharsetDecoder:
https://docs.oracle.com/javase/8/docs/api/java/nio/charset/CharsetDecoder.html

[3] Source code of Files.newBufferedReader(Path, Charset):
http://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/be30cb2a3088/src/share/classes/java/nio/file/Files.java#l2780

[4] Example of a file with non-UTF-8 characters inside:
https://github.com/CS2103JAN2018-F11-B4/main/blob/master/collated/functional/ZhangYijiong.md

[5] Source code of StreamDecoder setting malformed input to be replaced:
http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/sun/nio/cs/StreamDecoder.java#l230
```

### CODE REVIEW

**Comment by yong24s** [comment]
Tested. Works on non-UTF8 files.
Some nits on proposed message, else LGTM.

FileInfoExtractor#generateFileInfo(String, String) BufferedReader
only accepts and decodes UTF-8 character set~s~. However, some files
may contain special characters from ~an~other character sets[1], such as
ISO_8859_1.

This leads to an incomplete FileInfo being generated, in turn causing
the program to crash when analys~is~ing the file in 
FileInfoAnalyzer#aggregateBlameAuthorInfo(RepoConfiguration, FileInfo)
as the number of lines analyzed in git blame is greater than the number
of lines inside FileInfo.

Let's update the BufferedReader in
FileInfoExtractor#generateFileInfo(String, String) to accept and decode
ISO_8859_1 character set~s~.

[1]: Example of a file with non-UTF-8 characters inside:
https://github.com/CS2103JAN2018-F11-B4/main/blob/master/collated/functional/ZhangYijiong.md

**Comment by yamidark** [comment]
@eugenepeh For your review.

**Comment by eugenepeh** [comment]
@yong24s to properly record this, can you post the stack trace here as well?

File: src/functional/resources/dateRange/expected/reposense_testrepo-Beta/authorship.json [hunk 1]
```diff
@@ -25891,7 +25891,7 @@
         "author": {
           "gitId": "nbriannl"
         },
-        "content": "            + \": Opens a Google Maps view of a person’s address.\\n\""
+        "content": "            + \": Opens a Google Maps view of a personâs address.\\n\""
```
Comment (eugenepeh) [inline src/functional/resources/dateRange/expected/reposense_testrepo-Beta/authorship.json]: doesn't seems to be the optimal solution from the result of the content. Can you look into it further?

**Comment by yamidark** [comment]
@eugenepeh For your review again.

**Comment by yamidark** [comment]
@eugenepeh We probably need to add/replace 1 of the problematic repos with this error to the `testrepo`  for regressive testing.

**Comment by yong24s** [comment]
Some nits on the proposed message, else LGTM.

FileInfoExtractor#generateFileInfo(String, String)
and FileInfoAnalyzer#isReused(String,String) use~s~
Files#newBufferedReader(Path) to construct the BufferedReader for
reading of files.

~This BufferedReader~ The constructed **BufferedReader** uses **a**
default **UTF-8** Charset[1] and **a** default CharsetDecoder[2] for
decoding of the bytes read from the file into Strings.[3]

As some files being analyzed may contain special characters that is not
from the UTF-8 character set[3], this will cause the BufferedReader to
throw a java.nio.charset.MalformedInputException when it detects these
special characters.

As a result, an incomplete FileInfo is generated in
FileInfoExtractor#generateFileInfo(String, String)**, <---** which causes the
program to crash when analysing the file in 
FileInfoAnalyzer#aggregateBlameAuthorInfo(RepoConfiguration, FileInfo),
as the number of lines analyzed in git blame is greater than the number
of lines inside FileInfo. It will also cause the analysis of some
white-listed files to be skipped inside
FileInfoAnalyzer#isReused(String,String).

**Comment by eugenepeh** [comment]
>@eugenepeh We probably also need to add/replace 1 of the problematic repos with this error to the testrepo for regression testing.

To not overload the test, can you identify which is the file that cause this error?

**Comment by yamidark** [comment]
>To not overload the test, can you identify which is the file that cause this error?

Mentioned in the proposed message:
https://github.com/CS2103JAN2018-F11-B4/main/blob/master/collated/functional/ZhangYijiong.md

Have added a unit test for this file too.

File: src/main/java/reposense/authorship/FileInfoExtractor.java [hunk 2]
```diff
@@ -74,7 +76,7 @@ private static void getAllFileInfo(
     public static FileInfo generateFileInfo(String repoRoot, String relativePath) {
         FileInfo fileInfo = new FileInfo(relativePath);
         Path path = Paths.get(repoRoot, relativePath);
-        try (BufferedReader br = new BufferedReader(Files.newBufferedReader(path))) {
+        try (BufferedReader br = new BufferedReader(new InputStreamReader(new FileInputStream(path.toFile())))) {
```
[the remark this answers is not in the document]
Comment (yamidark) [reply src/main/java/reposense/authorship/FileInfoExtractor.java]: Yep, tested and it works :+1:, will convert it over since this is shorter and cleaner.

**Comment by eugenepeh** [comment]
Minor nit in proposed message, can further improve if necessary.

FileInfoExtractor#generateFileInfo(String, String)
and FileInfoAnalyzer#isReused(String,String) use
Files#newBufferedReader(Path) to construct the BufferedReader for
reading of files.

~The~ `This` constructed BufferedReader `uses` the default UTF-8 Charset[1]
and a default CharsetDecoder[2] ~for decoding of~ `to decode` the bytes
read from ~the~ `files` into Strings.[3]

As some files being analyzed may contain special characters that ~is~ `are` not
from the UTF-8 character set[3], ~this will~ `it causes` the BufferedReader to
throw a java.nio.charset.MalformedInputException.~ when it detects these
special characters.~

As a result, `it creates` an incomplete FileInfo ~is generated~ in `the`
FileInfoExtractor#generateFileInfo(String, String) `process`, which causes the
program to crash when analysing the file in 
FileInfoAnalyzer#aggregateBlameAuthorInfo(RepoConfiguration, FileInfo)~,~`. This is due to`
~as~ the number of lines analyzed in git blame ~is~ `being` greater than the number
of lines inside FileInfo. It will also cause the analysis of some
white-listed files to be skipped inside
FileInfoAnalyzer#isReused(String,String).

Let's update FileInfoExtractor#generateFileInfo(String, String)
and FileInfoAnalyzer#isReused(String,String) to construct the
BufferedReader with an java.io.FileReader instead.

The java.io.FileReader extends java.io.InputStreamReader,
which use the same default UTF-8 Charset. However, it will call 
StreamDecoder#forInputStreamReader(InputStream, Object, CharsetDecoder)
, which will set the CharsetDecoder to replace the malformed input
instead of throwing an error[5], thus allowing the FileInfo to be
constructed with the correct number of lines.

[1] Javadocs of Charset:
https://docs.oracle.com/javase/8/docs/api/java/nio/charset/Charset.html

[2] Javadocs of CharsetDecoder:
https://docs.oracle.com/javase/8/docs/api/java/nio/charset/CharsetDecoder.html

[3] Source code of Files.newBufferedReader(Path, Charset):
http://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/be30cb2a3088/src/share/classes/java/nio/file/Files.java#l2780

[4] Example of a file with non-UTF-8 characters inside:
https://github.com/CS2103JAN2018-F11-B4/main/blob/master/collated/functional/ZhangYijiong.md

[5] Source code of StreamDecoder setting malformed input to be replaced:
http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/sun/nio/cs/StreamDecoder.java#l230

**Comment by yamidark** [comment]
@damithc Prof for your review.

**Pull request merged**